packages/react/src/components/Dialog/Dialog.tsx

```diff
@@ -203,6 +204,20 @@ class Dialog extends AutoControlledComponent<WithAsProp<DialogProps>, DialogStat
     }
   }
 
+  handleDocumentKeydown = (getRefs: Function) => e => {
+    // if focus was lost from Dialog, for e.g. when click on Dialog's content
+    // and ESC is pressed, the opened Dialog should get closed and the trigger should get focus
+    const lastOverlayRef = getRefs().pop()
+    const isLastOpenedDialog: boolean =
+      lastOverlayRef && lastOverlayRef.current === this.overlayRef.current
+
+    const keyCode = keyboardKey.getCode(e)
+    if (keyCode === keyboardKey.Escape && isLastOpenedDialog && this.state.open) {
```

Comment:
```suggestion
    if (keyboardKey.getCode(e) === keyboardKey.Escape && isLastOpenedDialog) {
```

You don't need `this.state.open` as this event listener will be mounted only when `Dialog` is `open`:

```tsx
<Portal open={open} /* here we are */>
  <EventListener type="keydown" />
</Portal>
```